client/libra-dev/src/transaction.rs

```diff
@@ -0,0 +1,633 @@
+    transaction::{
+        helpers::TransactionSigner, RawTransaction, TransactionArgument, TransactionPayload,
+    },
+};
+use std::{convert::TryFrom, slice, time::Duration};
+use transaction_builder::{encode_transfer_script, get_transaction_name};
+
+#[no_mangle]
+pub unsafe extern "C" fn libra_SignedTransactionBytes_from(
+    sender_private_key_bytes: *const u8,
+    receiver: *const u8,
+    sequence: u64,
+    num_coins: u64,
+    max_gas_amount: u64,
+    gas_unit_price: u64,
+    expiration_time_secs: u64,
+    ptr_buf: *mut *mut u8,
+    ptr_len: *mut usize,
+) -> LibraStatus {
+    clear_error();
+    let private_key_buf: &[u8] =
+        slice::from_raw_parts(sender_private_key_bytes, ED25519_PRIVATE_KEY_LENGTH);
+    let private_key = match Ed25519PrivateKey::try_from(private_key_buf) {
+        Ok(result) => result,
+        Err(e) => {
+            update_last_error(format!("Invalid private key bytes: {}", e.to_string()));
+            return LibraStatus::InvalidArgument;
+        }
+    };
+
+    let public_key: Ed25519PublicKey = (&private_key).into();
+    let sender_address = AccountAddress::from_public_key(&public_key);
+
+    let receiver_buf = slice::from_raw_parts(receiver, ADDRESS_LENGTH);
+    let receiver_address = match AccountAddress::try_from(receiver_buf) {
+        Ok(result) => result,
+        Err(e) => {
+            update_last_error(format!("Invalid receiver address: {}", e.to_string()));
+            return LibraStatus::InvalidArgument;
+        }
+    };
+
+    let expiration_time = Duration::from_secs(expiration_time_secs);
+
+    let program = encode_transfer_script(&receiver_address, num_coins);
+    let payload = TransactionPayload::Script(program);
+    let raw_txn = RawTransaction::new(
+        sender_address,
+        sequence,
+        payload,
+        max_gas_amount,
+        gas_unit_price,
+        expiration_time,
+    );
+
+    let keypair = KeyPair::from(private_key);
+    let signer: Box<&dyn TransactionSigner> = Box::new(&keypair);
+    let signed_txn = match signer.sign_txn(raw_txn) {
+        Ok(result) => result,
+        Err(e) => {
+            update_last_error(format!("Error signing transaction: {}", e.to_string()));
+            return LibraStatus::InvalidArgument;
+        }
+    };
+
+    let signed_txn_bytes = match to_bytes(&signed_txn) {
+        Ok(result) => result,
+        Err(e) => {
+            update_last_error(format!(
+                "Error serializing signed transaction: {}",
+                e.to_string()
+            ));
+            return LibraStatus::InternalError;
+        }
+    };
+    let txn_buf: *mut u8 = libc::malloc(signed_txn_bytes.len()).cast();
+    txn_buf.copy_from(signed_txn_bytes.as_ptr(), signed_txn_bytes.len());
+
+    *ptr_buf = txn_buf;
+    *ptr_len = signed_txn_bytes.len();
+
+    LibraStatus::OK
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn libra_free_bytes_buffer(buf: *const u8) {
+    if !buf.is_null() {
```

Comment:
Should we be guarding this? If it is an error to call `libra_free_bytes_buffer` on a null pointer, this will hide the mistake. Perhaps we should assert instead.